File: JXSX_COD/githubDemo/VB_TCP连接示例/VB_TCP/VB_TCP/Form1.vb
Imports System.Net.Sockets
'使用到TcpListen类
Imports System.IO
'使用到StreamWriter类
Imports System.Net
'使用IPAddress类、IPHostEntry类等
Imports System.Threading

Public Class Form1
    Private swWriter As StreamWriter
    Private srRead As StreamReader
    '用以向网络基础数据流传送数据
    Private nsStream As NetworkStream
    '创建发送数据的网络基础数据流
    Private tcpClient As TcpClient
    '通过它实现向远程主机提出TCP连接申请
    Private tcpConnect As Boolean = False
    '定义标识符，用以表示TCP连接是否建立
    Private Sub butconnect_Click(sender As Object, e As EventArgs) Handles butconnect.Click
        Dim ipRemote As IPAddress
        'Dim tcpClient As TcpClient
        Try
            ipRemote = IPAddress.Parse(txtip.Text)
        Catch
            MessageBox.Show("输入的IP地址不合法！", "错误提示！")
            Return
            '判断给定的IP地址的合法性
        End Try
        Try
            tcpClient = New TcpClient(txtip.Text, Integer.Parse(txtport.Text))
            '对远程主机的8000端口提出TCP连接申请
            nsStream = tcpClient.GetStream()
            '通过申请，并获取传送数据的网络基础数据流
            swWriter = New StreamWriter(nsStream)
            srRead = New StreamReader(nsStream)
            '使用获取的网络基础数据流来初始化StreamWriter实例
            butconnect.Enabled = False
            butsend.Enabled = True
            butdis.Enabled = True
            tcpConnect = True
            txtmsg.AppendText("已经连接")

            Dim workerThread As New Thread(New ThreadStart(AddressOf DoWork))
            ' 开始执行线程  
            workerThread.IsBackground = True
            workerThread.Start()

            ' 等待工作线程完成  
            'workerThread.Join()

        Catch ex As Exception
            MessageBox.Show("无法和远程主机端口建立连接！", "错误提示！")
            Return
        End Try
    End Sub

    Private Sub butsend_Click(sender As Object, e As EventArgs) Handles butsend.Click
        swWriter.WriteLine(txtsend.Text)
        txtmsg.AppendText(txtsend.Text)
        swWriter.Flush()
    End Sub

    Private Sub DoWork()
        Dim receiveMessage As String = String.Empty
        ''''''''''''''''''''''''''''''''''''''''''''''''
        Try
            While (tcpConnect)
                While (True)
                    receiveMessage = srRead.ReadLine()
                    If String.IsNullOrEmpty(receiveMessage) Then
                        Continue While
                    End If
                    Exit While
                End While


                If (String.IsNullOrEmpty(receiveMessage)) Then
                    If Not tcpConnect Then
                        srRead.Close()
                        tcpClient.Close()
                        Exit While
                    End If
                    Continue While
                End If

                'If String.IsNullOrEmpty(receiveMessage.Trim()) Then
                '    Thread.Sleep(100)
                '    Continue While
                'End If
                'txtmsg.AppendText(receiveMessage)
                'ShowMsg(receiveMessage)

                'Me.Invoke(Sub() txtmsg.AppendText(receiveMessage))
                Me.Invoke(New VoidDelegate(AddressOf UpdateText), receiveMessage)
            End While
        Catch ex As Exception
            'MsgBox("提示")
            MessageBox.Show("无法和远程主机端口建立连接！", "错误提示！")
        End Try
        ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''



        'While (Not tcpConnect)
        '    '客户端主动断开连接前检查流中是否还是数据
        '        If (tcpClient.Available <= 0) Then
        '        If (Not IsOnline(tcpClient)) Then
        '            Exit While
        '        End If
        '        Continue While
        '    End If

        '    'if (client.Client == null) return;

        '    Dim d As New ReceiveMessageDelegate(AddressOf receiveMessage), receiveString)

        '    'ReceiveMessageDelegate d = New ReceiveMessageDelegate(receiveMessage)
        '        IAsyncResult result = d.BeginInvoke(out receiveString, null, null)
        '    '使用轮询方式来判断异步操作是否完成
        '    While (Not result.IsCompleted)
        '        If (tcpConnect) Then
        '            Exit While
        '        End If
        '            Thread.Sleep(50);
        '    End While
        '    '获取Begin方法的返回值和所有输入/输出参数
        '        d.EndInvoke(out receiveString, result)
        '    If (String.IsNullOrEmpty(receiveString)) Then
        '        If (Not tcpConnect) Then
        '            tcpClient.Close()
        '            swWriter.Close()
        '            srRead.Close()
        '            Exit While
        '        End If
        '        Continue While
        '    End If
        'End While


    End Sub

    Private Function IsOnline(ByRef c As TcpClient) As Boolean
        Try
            Return Not (c.Client.Poll(1000, SelectMode.SelectRead) And c.Client.Available = 0 Or c.Client Is Nothing)
        Catch
            Return False
        End Try
    End Function

    Public Delegate Sub VoidDelegate(ByVal txtmsg As String)
    Public Delegate Sub ReceiveMessageDelegate(ByRef receiveMessage As String)

    Public Sub UpdateText(ByVal receiveMessage As String)
        txtmsg.AppendText(receiveMessage)
    End Sub

    Private Sub butdis_Click(sender As Object, e As EventArgs) Handles butdis.Click

        If tcpConnect Then
            tcpConnect = False
            tcpClient.Close()
            swWriter.Close()
            srRead.Close()
            butconnect.Enabled = True
            butdis.Enabled = False
            butsend.Enabled = False
        End If

    End Sub
End Class

File: JXSX_COD/C#网络编程/TCPclient异步执行/TCPclient/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.IO;

namespace TCPclient
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
        TcpClient client;
        StreamReader sr;
        StreamWriter sw;
        string ip = string.Empty;
        int port = 0;
        private bool isExit = false;
        private void butconnect_Click(object sender, EventArgs e)
        {
            isExit = false;
            string[] ipport = new string[2];
            ipport = txtIP.Text.Split(':');
            ip = ipport[0];
            port = Convert.ToInt32(ipport[1]);
            IPAddress IP = IPAddress.Parse(ip);
            client = new TcpClient();
            IAsyncResult result = client.BeginConnect(ip, port, null, null);
            while (result.IsCompleted == false)
            {
                Thread.Sleep(100);
            }
            try
            {
                client.EndConnect(result);
                NetworkStream networkstream = client.GetStream();
                sr = new StreamReader(networkstream,Encoding.Default);    //可以读写中英文
                sw = new StreamWriter(networkstream, Encoding.Default);
                sendMsg("hello");
                Thread startRecevie = new Thread(recevie);
                startRecevie.IsBackground = true;
                startRecevie.Start();
                riTxt.Text = "连接成功";
                return;
            }
            catch
            {
                riTxt.Text = "连接失败";
                return;
            }
        }

        private delegate void sendmessDEL(string msg);
        private void sendMsg(string msg)
        {
            byte[] buffer = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x06, 0x01, 0x06, 0x00, 0x00, 0x00, 0x01 };
            sw.Write(buffer);
            //sw.WriteLine(msg);
            sw.Flush();
        }


        private delegate void recevieData(out string recevieMsg);
        private void recevie()
        {
            while (isExit == false)
            {
                //if (client.Available <= 0) continue;
                //if (client.Client == null) break;

                string recevieMsg = null;
                recevieData d = new recevieData(receiveDataMsg);
                IAsyncResult result = d.BeginInvoke(out recevieMsg, null, null);    //异步执行接收消息，BeginInvoke -- EndInvoke
                while (result.IsCompleted == false)
                {
                    Thread.Sleep(50);
                    if (isExit)
                    {
                        sr.Close();
                        sw.Close();
                        client.Close();
                        addriTxt("连接断开");
                        return;
                    }
                }
                d.EndInvoke(out recevieMsg, result);
                if (recevieMsg == null && client.Available <= 0)    //检测服务器主动断开连接
                {
                    isExit = true;
                    sr.Close();
                    sw.Close();
                    client.Close();
                    addriTxt("服务器断开");
                }
                addriTxt(recevieMsg);
            }
            return;
        }
        private void receiveDataMsg(out string recevieMsg)
        {
            recevieMsg = null;
            try
            {
                recevieMsg = sr.ReadLine();
                return;
            }
            catch
            {
                addriTxt("读取错误");
            }
        }

        private delegate void addriTxtDEL(string Msg);
        private void addriTxt(string Msg)    //跨线程访问控件
        {
            if (riTxt.InvokeRequired)
            {
                addriTxtDEL d = new addriTxtDEL(addriTxt);
                riTxt.Invoke(d, new object[] { Msg });
            }
            else
            {
                riTxt.AppendText(Msg + Environment.NewLine);
                //this.Invoke((EventHandler)(delegate { riTxt.AppendText(Msg); }));
            }
        }

        private void butdiscon_Click(object sender, EventArgs e)
        {
            isExit = true;
            ritxtsend.Text = "客户端断开";
            butsend_Click(null, null);
        }

        private void butsend_Click(object sender, EventArgs e)
        {
            string msg = ritxtsend.Text;
            try
            {
                if(!isExit)
                {
                    sendmessDEL d = new sendmessDEL(sendMsg);
                    IAsyncResult result = d.BeginInvoke(msg, null, null);   //异步接收消息
                    while (result.IsCompleted == false)
                    {
                        Thread.Sleep(50);
                    }
                    status st = new status();
                    st.d = d;
                    st.result = result;
                    Thread t = new Thread(endinvoke);
                    t.IsBackground = true;
                    t.Start(st);   //只能有一个参数
                }
            }
            catch
            {
                addriTxt("主动发送失败");
            }
        }

        private struct status
        {
            public sendmessDEL d;
            public IAsyncResult result;
        }
        private void endinvoke(object st)    //Thread.Start只能有一个参数，所有传参结构体
        {
            status sta = (status)st;
            sta.d.EndInvoke(sta.result);   //EndInvoke阻塞线程，所以需要采用异步执行
            return;
        }

        private void Form1_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (client.Client != null)
            {
                Thread.Sleep(10);
                sr.Close();
                sw.Close();
                client.Close();
            }
        }
    }
}
